odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2385 @@
+            return real_space.element(self.data.imag)
+
+    @imag.setter
+    def imag(self, newimag):
+        """Setter for the imaginary part.
+
+        This method is invoked by ``x.imag = other``.
+
+        Parameters
+        ----------
+        newimag : array-like or scalar
+            Values to be assigned to the imaginary part of this element.
+
+        Raises
+        ------
+        ValueError
+            if the space is real, i.e., no imagninary part can be set
+
+        Examples
+        --------
+        >>> space = odl.cn((2, 3))
+        >>> x = space.element([[1 + 1j, 2, 3 - 3j],
+        ...                    [4, 5 - 5j, 6]])
+        >>> zero = odl.rn((2, 3)).zero()
+        >>> x.imag = zero
+        >>> x
+        cn((2, 3)).element(
+            [[(1+0j), (2+0j), (3+0j)],
+             [(4+0j), (5+0j), (6+0j)]]
+        )
+
+        Other array-like types and broadcasting:
+
+        >>> x.imag = 1.0
+        >>> x
+        cn((2, 3)).element(
+            [[(1+1j), (2+1j), (3+1j)],
+             [(4+1j), (5+1j), (6+1j)]]
+        )
+        >>> x.imag = [[2, 3, 4],
+        ...           [5, 6, 7]]
+        >>> x
+        cn((2, 3)).element(
+            [[(1+2j), (2+3j), (3+4j)],
+             [(4+5j), (5+6j), (6+7j)]]
+        )
+        """
+        if self.space.is_real_space:
+            raise ValueError('cannot set imaginary part in real spaces')
+        self.imag.data[:] = newimag
+
+    def conj(self, out=None):
+        """Return the complex conjugate of ``self``.
+
+        Parameters
+        ----------
+        out : `NumpyTensor`, optional
+            Element to which the complex conjugate is written.
+            Must be an element of ``self.space``.
+
+        Returns
+        -------
+        out : `NumpyTensor`
+            The complex conjugate element. If ``out`` was provided,
+            the returned object is a reference to it.
+
+        Examples
+        --------
+        >>> space = odl.cn((2, 3))
+        >>> x = space.element([[1 + 1j, 2, 3 - 3j],
+        ...                    [4, 5 - 5j, 6]])
+        >>> x.conj()
+        cn((2, 3)).element(
+            [[(1-1j), (2-0j), (3+3j)],
+             [(4-0j), (5+5j), (6-0j)]]
+        )
+
+        The out parameter allows you to avoid a copy
+
+        >>> y = space.element()
+        >>> y_out = x.conj(out=y)
+        >>> y
+        cn((2, 3)).element(
+            [[(1-1j), (2-0j), (3+3j)],
+             [(4-0j), (5+5j), (6-0j)]]
+        )
+        >>> y_out is y
+        True
+
+        It can also be used for in-place conjugation:
+
+        >>> x_out = x.conj(out=x)
+        >>> x
+        cn((2, 3)).element(
+            [[(1-1j), (2-0j), (3+3j)],
+             [(4-0j), (5+5j), (6-0j)]]
+        )
+        """
+        if out is None:
+            return self.space.element(self.data.conj())
+        else:
+            self.data.conj(out.data)
+            return out
+
+    def __getitem__(self, indices):
```

Comment:
Looked again, now I see why. Anyway my suggestion for these is that you actually mirror the ideas of `ProductSpace` where we also add `__getitem__` to the space [see here](https://github.com/odlgroup/odl/blob/5432dfc050485cb0d5f3befa3c1fccb97771f976/odl/space/pspace.py#L565-L574).

Anyway, this would allow you to remove this and simply change this function (in `NumpyGeneralizedTensor`) to

```python
arr = self.data[indices]
return self.space[indices].element(arr)
```